### ISSUE TITLE
[MNT] how to deal with problematic base class imports

### ISSUE THREAD

**Issue by MatthewMiddlehurst** [sktime/sktime#3660]
There seems to be no structure to the placement of individual package base classes (or if there is, it is very broad). Most are situated in the main package folder i.e. `sktime.classification.base.py`, but the files surrounding the base classes and their availability in the `__init__.py` file will change from package to package. 

For base classes which are in an `__init__.py` with other estimators or situated in a subpackage, any imports will also import the surrounding estimators. While rather niche, this can possibly lead to situations like the one seen in #3634 where unrelated estimators are being imported through base class imports in the `registry` package.

I think it would be better to have a similar structure for all packages.

A few ideas, though feel free to suggest others:

1. Move base classes to the `base` package, i.e. `from sktime.base import BaseForecaster` or `from sktime.base._base import BaseForecaster`. Possibly in subpackages i.e. `from sktime.base.forecasting import BaseForecaster` or `from sktime.base.forecasting._base import BaseForecaster`.
2. Place the files in the main package i.e. `from sktime.forecasting import BaseForecaster` or `from sktime.forecasting._base import BaseForecaster`. This would require no files other than the base class to use the top level `__init__.py`.
3. Have base class subpackages i.e. `from sktime.forecasting.base import BaseForecaster` or `from sktime.forecasting.base._base import BaseForecaster`. This also requires no top level `__init__.py` usage.
4. Do nothing.

**Comment by fkiraly** [comment]
disagree with any of that. #3634 is just a case of sloppy imports imo.

The current way forward is the `skbase` package led by @RNKuhns (formerly `BaseObject`).

**Comment by MatthewMiddlehurst** [comment]
So are `BaseForecaster`, `BaseClassifier`, `BaseEarlyClassifier` etc. going to move to `skbase`?

**Comment by fkiraly** [comment]
no

**Comment by MatthewMiddlehurst** [comment]
I don't really see how this issue and that conflict then, if these base classes are present (using the current contents or `skbase`) they have to be placed somewhere.

**Comment by fkiraly** [comment]
the import `from sktime.submodule.base import BaseClassName` should work consistently.

It is imo bad design to try moving the base classes to a joint base package or future `skbase`.
https://en.wikipedia.org/wiki/Coupling_(computer_programming)
because that would lead to high coupling of all the submodules with each other.

**Comment by MatthewMiddlehurst** [comment]
Yes I agree, I prefer either 2 or 3 from the main issue. The problem is the way some of the packages are organised, these import issues through the base class are appearing.

**Comment by fkiraly** [comment]
> I'm not quite sure how putting a new directory in the root breaks things, but if it does, then unfortunately this method will not work.

To answer this: it would break cohesion/coupling patterns, and  high-level architectural patterns for modules (they should "look the same"). Also, it looks like a deprecation nightmare.

**Comment by fkiraly** [comment]
> this can possibly lead to situations like the one seen in https://github.com/sktime/sktime/pull/3634 where unrelated estimators are being imported through base class imports in the registry package.

Can you please show a worked example?

I have been trying to understand this, and the estimators in #3634 seem to import directly from `_base` in the distances module rather than from the module level `__init__`, so why would this import adjacent estimators too?

**Comment by MatthewMiddlehurst** [comment]
As I commented in the PR, our test case is this code with no version of `numba` installed (to mimic the problem mentioned by @ngupta23):

```
from sklearn.preprocessing import MinMaxScaler
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import ForecastingPipeline
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.naive import NaiveForecaster
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.impute import Imputer

if __name__ == "__main__":
    y, X = load_longley()
    y_train, _, X_train, X_test = temporal_train_test_split(y, X)
    fh = ForecastingHorizon(X_test.index, is_relative=False)
    pipe = ForecastingPipeline([
        Imputer(method="mean"),
        TabularToSeriesAdaptor(MinMaxScaler()),
        NaiveForecaster(strategy="drift"),
    ])
    pipe.fit(y_train, X_train)
    y_pred = pipe.predict(fh=fh, X=X_test)
    print(y_pred)
```

In #3634 we tested a few methods before reaching the monster we ended up at. 

If we import using this in the registry:
```
from sktime.dists_kernels._base import (
    BasePairwiseTransformer,
    BasePairwiseTransformerPanel,
)
```
We get the error:
```
Traceback (most recent call last):
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\_contrib\local_code2.py", line 4, in <module>
    from sktime.forecasting.compose import ForecastingPipeline
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\forecasting\compose\__init__.py", line 36, in <module>
    from sktime.forecasting.compose._pipeline import (
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\forecasting\compose\_pipeline.py", line 12, in <module>
    from sktime.registry import scitype
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\registry\__init__.py", line 4, in <module>
    from sktime.registry._base_classes import (
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\registry\_base_classes.py", line 61, in <module>
    from sktime.dists_kernels._base import (
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\dists_kernels\__init__.py", line 10, in <module>
    from sktime.dists_kernels.compose_tab_to_panel import AggrDist, FlatDist
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\dists_kernels\compose_tab_to_panel.py", line 15, in <module>
    import sktime.dists_kernels.distances.compose_tab_to_panel as new_class_loc
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\dists_kernels\distances\__init__.py", line 5, in <module>
    from sktime.dists_kernels.distances.dtw import DtwDist
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\dists_kernels\distances\dtw.py", line 10, in <module>
    from sktime.distances import pairwise_distance
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\distances\__init__.py", line 33, in <module>
    from sktime.distances._distance import (
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\distances\_distance.py", line 8, in <module>
    from numba import njit
ModuleNotFoundError: No module named 'numba'

Process finished with exit code 1
```
a) we can't import these classes using `submodule.base`, only `submodule._base` and `submodule`
b) we are still importing from dists_kernels `__init__.py`: 
```
  File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\dists_kernels\__init__.py", line 10, in <module>
    from sktime.dists_kernels.compose_tab_to_panel import AggrDist, FlatDist
```

If we were to move it away from the `__init__.py` into another directory, i.e.  `sktime.dists_kernels.base._base`, we would get the same error, except for this bit showing us importing from the new dir:   
```
File "D:\CMP Machine Learning\sktime-workshop-boss\sktime\registry\_base_classes.py", line 61, in <module>
    from sktime.dists_kernels.base._base import (
```

This leaves us in a bit of a pickle. We can't import anywhere from `dists_kernels` currently without also importing the distances from `__init__.py` and whatever they have with them. Fixing this would require removing them from the `__init__.py` which would require deprecation. Therefore, the only immediate fix (to the import issue) is to not import from `dists_kernels` temporarily.

**Comment by fkiraly** [comment]
I'm still confused.

```python
from sktime.dists_kernels._base import (
    BasePairwiseTransformer,
    BasePairwiseTransformerPanel,
)
```

works for me with `numba` installed in the env.

Without `numba` in the env it does not work, but that is expected since `numba` is currently not isolated (as a core dependency), and the distance classes import the numba distances.

Can you be more explicit and precise here?

1. given what condition
2. given what command executed how
3. what happens
4. why should this not happen
5. what should happen instead

?

**Comment by MatthewMiddlehurst** [comment]
The fact the distance classes are being imported from a completely unrelated function through the registry is the issue.

If you do not think that is an issue, feel free to close this.

**Comment by fkiraly** [comment]
> If you do not think that is an issue, feel free to close this.

Seems like an issue, may I kindly ask which import this happens for, please?

1. given what condition
2. given what command executed how
3. what happens
4. why should this not happen
5. what should happen instead

?

**Comment by MatthewMiddlehurst** [comment]
The imports happen through `ForecastingPipeline`, but the numba issue is just an example of the imports happening because it's easy to see which changes stop the imports (the error stops happening). 

If other classes such as distances and early classifiers being imported with the base classes is fine, we can just close this issue.

**Comment by fkiraly** [comment]
Still don't get it - I would really appreciate a step-by-step answer to my questions, with code examples.

**Comment by MatthewMiddlehurst** [comment]
I'm sorry, I don't really know what to give past what I already have. What kinda of code example would you like to display that when you import `ForecastingPipeline` you are also importing `EditDist` or whatever else. It is more of a design issue.

1. `ForecastingPipeline` has the import
`from sktime.registry import scitype`
 
2. which imports all base classes
`from sktime.registry._base_classes`

3. one of which is the distances base
  ```
  from sktime.dists_kernels._base import (
      BasePairwiseTransformer,
      BasePairwiseTransformerPanel,
  )
 ```
4.  through which most of the distances classes are imported
`from sktime.dists_kernels.compose_tab_to_panel import AggrDist, FlatDist`

Question: Should the distance base class be coupled with the other distance classes such they are all imported when you import any of them.

`ForecastingPipeline` needs to enter the registry, but does it also need to import the distances and compile all the numba functions in the distances package (just an example, it doesn't matter whether numba is installed, uninstalled or a soft dependency).

**Comment by fkiraly** [comment]
thanks, this is much clearer, since it actually gives an example and the mechanism is spelled out.

I think there are two issues:

* a concrete issue, where `_pipeline.py` imports `scitype`, which is more of a testing utility and coupled to all base classes. This import probably should not happen in estimators, and it seems the pipelines are the only non-test/utility case where it is imported. This can be easily fixed.
* there is the more general question whether we want to couple the `registry` module to all the base classes in this way - perhaps not? I think that's the main problem rather than where the base classes live, at least it is a possible way to address the issue - change the import logic in `registry`, for instance, rather than moving all the base classes around.

**Comment by fkiraly** [comment]
This should fix the coupling for the forecasting pipeline:
https://github.com/sktime/sktime/pull/3799

Are there any other instances where this behaviour happens?

A quick search for `from registry` indicates this might be the only example? Or are you aware of others?

**Comment by fkiraly** [comment]
fixed in #6970
